src/resolvers.ts

```diff
@@ -0,0 +1,130 @@
+import { Book} from "./models/book";
+import { Customer } from "./models/customer";
+import { LendingRecord } from "./models/lendingRecord";
+import { BooksDataSource } from "./dataSources/booksDataSource";
+import { CustomersDataSource } from "./dataSources/customersDataSource";
+import { logger } from "./logger";
+
+const booksDataSource = new BooksDataSource();
+const customersDataSource = new CustomersDataSource();
+
+export const resolvers = {
+  Query: {
+    // Get all books
+    books: (): Book[] => booksDataSource.getBooks(),
+
+    // Get a book by id
+    book: (_: any, { id }: { id: string }): Book | undefined => {
+      return booksDataSource.getBookById(id);
+    },
+
+    // Get all customers
+    customers: (): Customer[] => customersDataSource.getCustomers(),
+
+    // Get a customer by id
+    customer: (_: any, { id }: { id: string }): Customer | undefined =>
+      customersDataSource.getCustomerById(id),
+  },
+
+  Mutation: {
+    // Add a new book, the isLent field is set to false by default
+    addBook: (
+      _: any,
+      {title, author }: {title: string; author: string }
+    ): Book => {
+      const newBook = new Book(title, author);
+      booksDataSource.addBook(newBook);
+      return newBook;
+    },
+
+    // Add a new customer
+    addCustomer: (
+      _: any,
+      {name, email }: {name: string; email: string }
+    ): Customer => {
+      const newCustomer = new Customer(name, email);
+      customersDataSource.addCustomer(newCustomer);
+      return newCustomer;
+    },
+
+    // Lend a book to a customer, updates the fields of a book related to lending
+    lendBook: (
+      _: any,
+      {
+        bookId,
+        customerId,
+        lentDate,
+        dueDate,
+      }: {
+        bookId: string;
+        customerId: string;
+        lentDate: string;
+        dueDate: string;
+      }
+    ): Book => {
+      const book = booksDataSource.getBookById(bookId);
+      const customer = customersDataSource.getCustomerById(customerId);
+
+      if (book && customer && !book.isLent) {
```

Comment:
I highly suggest you read this [blog post](https://levelup.gitconnected.com/follow-these-6-patterns-or-i-will-reject-your-pull-request-fc08f908e7fe).

What we tend to do in order make our code more readable is use early returns. You can inverse your logic here and check for the bad paths first and throw the appropriate errors. This is called keeping the happy path on the left side.

Reply:
I read the article and it was very insightful. Now I refactored the code to follow this early return principle. This allowed for better error messages and is also easier to understand. Thanks!